community/index/src/main/java/org/neo4j/index/internal/gbptree/TreeNodeDynamicSize.java

```diff
@@ -19,44 +19,188 @@
+import static org.neo4j.index.internal.gbptree.TreeNode.Type.LEAF;
+
 public class TreeNodeDynamicSize<KEY, VALUE> extends TreeNode<KEY,VALUE>
 {
+    private static final int BYTE_POS_ALLOCSPACE = BASE_HEADER_LENGTH;
+    private static final int HEADER_LENGTH_DYNAMIC = BYTE_POS_ALLOCSPACE + BYTE_SIZE_OFFSET;
+
+    private static final int LEAST_NUMBER_OF_ENTRIES_PER_PAGE = 2;
+    private static final int MINIMUM_ENTRY_SIZE_CAP = Long.SIZE;
+    private final int keyValueSizeCap;
+
     TreeNodeDynamicSize( int pageSize, Layout<KEY,VALUE> layout )
     {
         super( pageSize, layout );
+
+        keyValueSizeCap = (pageSize - HEADER_LENGTH_DYNAMIC) / LEAST_NUMBER_OF_ENTRIES_PER_PAGE - BYTE_SIZE_TOTAL_OVERHEAD;
+
+        if ( keyValueSizeCap < MINIMUM_ENTRY_SIZE_CAP )
+        {
+            throw new MetadataMismatchException(
+                    "We need to fit at least %d key-value entries per page in leaves. To do that a key-value entry can be at most %dB " +
+                            "with current page size of %dB. We require this cap to be at least %dB.",
+                    LEAST_NUMBER_OF_ENTRIES_PER_PAGE, keyValueSizeCap, pageSize, Long.SIZE );
+        }
+    }
+
+    @Override
+    void writeAdditionalHeader( PageCursor cursor )
+    {
+        setAllocSpace( cursor, pageSize );
+    }
+
+    private void setAllocSpace( PageCursor cursor, int allocSpace )
+    {
+        cursor.setOffset( BYTE_POS_ALLOCSPACE );
+        putKeyOffset( cursor, allocSpace );
+    }
+
+    int getAllocSpace( PageCursor cursor )
+    {
+        cursor.setOffset( BYTE_POS_ALLOCSPACE );
+        return readKeyOffset( cursor );
     }
 
     @Override
     KEY keyAt( PageCursor cursor, KEY into, int pos, Type type )
     {
-        throw new UnsupportedOperationException( "Implement me" );
+        placeCursorAtActualKey( cursor, pos, type );
+
+        // Read key
+        int keySize = readKeySize( cursor );
+        if ( keySize > keyValueSizeCap || keySize < 0 )
+        {
+            cursor.setCursorException( format( "Read unreliable key, keySize=%d, keyValueSizeCap=%d, keyHasTombstone=%b",
+                    keySize, keyValueSizeCap, hasTombstone( keySize ) ) );
+        }
+        if ( type == LEAF )
+        {
+            progressCursor( cursor, DynamicSizeUtil.BYTE_SIZE_VALUE_SIZE );
+        }
+        layout.readKey( cursor, into, keySize );
+        return into;
+    }
+
+    private void placeCursorAtActualKey( PageCursor cursor, int pos, Type type )
+    {
+        // Set cursor to correct place in offset array
+        int keyPosOffset = keyPosOffset( pos, type );
+        cursor.setOffset( keyPosOffset );
+
+        // Read actual offset to key
+        int keyOffset = readKeyOffset( cursor );
+
+        // Verify offset is reasonable
+        if ( keyOffset > pageSize )
+        {
+            cursor.setCursorException( "Tried to read key on offset " + keyOffset + ". Page size is " + pageSize );
```

Comment:
I think this was put in during development when there was actually bad keySizes around because of bad overwrites. And if bad keySizes are actually around we don't necessarily notice those (as long as they are non negative), because we are not necessarily out of bounds.